core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java

```diff
@@ -91,29 +98,38 @@ public void ensureCatalogsLoaded(Session session, List<CatalogProperties> expect
             return;
         }
 
-        catalogsUpdateLock.lock();
+        Lock catalogLoadLock = catalogsUpdateLock.readLock();
+        catalogLoadLock.lock();
         try {
             if (stopped) {
                 return;
             }
 
-            for (CatalogProperties catalog : getMissingCatalogs(expectedCatalogs)) {
-                checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
-                CatalogConnector newCatalog = catalogFactory.createCatalog(catalog);
-                catalogs.put(catalog.getCatalogHandle(), newCatalog);
-                log.debug("Added catalog: " + catalog.getCatalogHandle());
-            }
+            CompletableFuture.allOf(
+                getMissingCatalogs(expectedCatalogs).stream()
+                        .map(catalog ->
+                                CompletableFuture.runAsync(() -> {
+                                    checkArgument(!catalog.getCatalogHandle().equals(GlobalSystemConnector.CATALOG_HANDLE), "Global system catalog not registered");
+                                    catalogs.computeIfAbsent(catalog.getCatalogHandle(), ignore -> {
```

Comment:
What's the purpose of `computeIfAbsent` here? There shouldn't be any duplicated keys, so there's no point in trying to avoid computation.

Reply:
Compute if absent should allow two tasks/queries to load the same catalog at once, only one of them will do the work.

Reply:
The read lock guarding this section allows multiple loaders at once

Reply:
Got it. It used to be an exclusive lock that guaranteed all load operations across all queries were serialized.